libyul/optimiser/SyntacticalEquality.cpp

```diff
@@ -29,48 +29,143 @@ using namespace std;
 using namespace dev;
 using namespace yul;
 
-bool SyntacticalEqualityChecker::equal(Expression const& _e1, Expression const& _e2)
+bool SyntacticallyEqual::operator()(Expression const& _lhs, Expression const& _rhs)
 {
-	if (_e1.type() != _e2.type())
+	return boost::apply_visitor([this](auto&& _lhs, auto&& _rhs) -> bool {
+		// ``this->`` is redundant, but required to work around a bug present in gcc 6.x.
+		return this->expressionEqual(_lhs, _rhs);
+	}, _lhs, _rhs);
+}
+
+bool SyntacticallyEqual::operator()(Statement const& _lhs, Statement const& _rhs)
+{
+	return boost::apply_visitor([this](auto&& _lhs, auto&& _rhs) -> bool {
+		// ``this->`` is redundant, but required to work around a bug present in gcc 6.x.
+		return this->statementEqual(_lhs, _rhs);
+	}, _lhs, _rhs);
+}
+
+bool SyntacticallyEqual::expressionEqual(FunctionalInstruction const& _lhs, FunctionalInstruction const& _rhs)
+{
+	return
+		_lhs.instruction == _rhs.instruction &&
+		compareVector<Expression, &SyntacticallyEqual::operator()>(_lhs.arguments, _rhs.arguments)
+	;
+}
+
+bool SyntacticallyEqual::expressionEqual(FunctionCall const& _lhs, FunctionCall const& _rhs)
+{
+	return
+		expressionEqual(_lhs.functionName, _rhs.functionName) &&
+		compareVector<Expression, &SyntacticallyEqual::operator()>(_lhs.arguments, _rhs.arguments)
+	;
+}
+
+bool SyntacticallyEqual::expressionEqual(Identifier const& _lhs, Identifier const& _rhs)
+{
+	auto lhsIt = m_identifiersLHS.find(_lhs.name);
+	auto rhsIt = m_identifiersRHS.find(_rhs.name);
+	return
+		(lhsIt == m_identifiersLHS.end() && rhsIt == m_identifiersRHS.end() && _lhs.name == _rhs.name) ||
+		(lhsIt != m_identifiersLHS.end() && rhsIt != m_identifiersRHS.end() && lhsIt->second == rhsIt->second)
+	;
+}
+bool SyntacticallyEqual::expressionEqual(Literal const& _lhs, Literal const& _rhs)
+{
+	return _lhs.kind == _rhs.kind && _lhs.value == _rhs.value && _lhs.type == _rhs.type;
```

Comment:
Would probably keep it like this, but we could parse the value and compare the u256 value.

Reply:
I changed the comparison to literal values and it already has quite some effects on the optimization results. See commit number 4. Note that the function combiner itself is not activated at that point, but the changes are due to ``SyntacticallyEqual`` being used in the CSE and the simplification rules. In particular it's interesting that a function is inlined due to this change that wasn't before.